### PR TITLE
Headless tests

### PR DESCRIPTION
**Pull request by wolframkriesing**
This adds `npm test` that uses [puppeteer](https://github.com/GoogleChrome/puppeteer) for headless testing on Chrome. This allows a fast local feedback cycle, no need to open the browser. But it runs **only** in (headless) Chrome.

Updates to docs and README accordingly.

### CODE REVIEW

**Comment by robbear** [comment]
@wolframkriesing Thank you so much for the PR. @JanMiksovsky will have a look when he returns from holiday. One thing you might adjust: We try to nail down our package.json dependencies, so it would be better if you removed the hat-version SemVer values (eg "^1.0.0") with fixed values ("1.0.0").

**Comment by JanMiksovsky** [comment]
I'm looking at this now. This PR is very well done.

My main question is whether it's worth adding another way of testing to our toolchain. This doesn't appear to be noticeably faster than hitting Refresh in Chrome itself, and testing in the browser has the advantage of making it easy to move from identifying test failures to diagnosing them.

Can you share more about how you normally use a headless test like this, and what you see as the advantages? Do you normally have it rigged to run automatically when files change, for example? (Using LiveReload for that purpose in Chrome is slow, so I normally manually reload.)

File: test/index.html [hunk 1]
```diff
@@ -19,7 +19,11 @@
 
   <script type="module">
     import * as tests from './tests.js';
-    mocha.run();
+    let errors = [];
```
Comment (JanMiksovsky) [inline test/index.html]: Since `errors` is never redefined, this can be a `const`.

**Comment by wolframkriesing** [comment]
@JanMiksovsky that's a good point, which I was thinking about the last days too. In the first place, it's that I normally like to run the tests as close as possible to where I am coding, which normally is not the browser.
But I think the main thing for me is the speed up of the TDD cycle. I can code+safe and get the fastest feedback possible (within my IDE), without switching to the browser. The way I am used to refactor in the tiniest steps possible, I just safe and run tests, keep coding. That might be very subjective, but I just feel that tiny steps and fast feedback are essential, especially when the code and the number of tests will grow.
Also I am actually on the search for a pure command-line based way to test WebComponents and after chatting with a couple of people I think headless is the way forward. When multiple browsers come into play I would like to get feedback from all of them asap, via headless this seems possible.

I would see headless-Firefox be the next thing to implement, so there would be two browsers covered. Which would not be fun anymore to do by manual reload.

Anyway, I think this is just one way to validate the code. As long as there is no alternative to real in-browser tests that can be as complete as saucelabs-kinda services, we need to run all of them, to ensure working WebComponents.

**Comment by JanMiksovsky** [comment]
Okay, this sounds worth trying! I don't personally run a command line inside my IDE, but I can imagine many people do, and I'm willing it to try using that myself. We can observe how useful it proves to project maintainers in practice, and if we later conclude it's not actually being used, we can drop it.

Last remaining (minor) thoughts:
* Why port 3718 for `start`? Not that this makes a functional difference, I'm just curious. I've generally seen ports like 8080, 3000, 5000, etc. as defaults.
* I'd recommend moving headless-tests.js to the /tasks directory, as that directory contains all our other npm scripts (linting, building, etc.).

**Comment by wolframkriesing** [comment]
Will move the code when I get back online.
The port was actually my attempt to write elix in leet ;) but I don't care
about it.

cu
Wolfram

Am 05.01.2018 02:34 schrieb "Jan Miksovsky" <notifications@github.com>:

Okay, this sounds worth trying! I don't personally run a command line
inside my IDE, but I can imagine many people do, and I'm willing it to try
using that myself. We can observe how useful it proves to project
maintainers in practice, and if we later conclude it's not actually being
used, we can drop it.

Last remaining (minor) thoughts:

   - Why port 3718 for start? Not that this makes a functional difference,
   I'm just curious. I've generally seen ports like 8080, 3000, 5000, etc. as
   defaults.
   - I'd recommend moving headless-tests.js to the /tasks directory, as
   that directory contains all our other npm scripts (linting, building, etc.).

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/elix/elix/pull/55#issuecomment-355449233>, or mute the
thread

**Comment by JanMiksovsky** [comment]
> The port was actually my attempt to write elix in leet ;) but I don't care
about it.

That's hilarious. I actually did try to read it as leet, but couldn't figure out the 8!

Just to pick something simple, let's use 3000.

**Comment by JanMiksovsky** [comment]
Thanks for this PR!

**Pull request merged**